internal/yaml/parser_test.go

```diff
@@ -0,0 +1,221 @@
+// package yaml offers a simplified YAML parser abstraction that not only parses YAML into structures, but also offers
+// access to the YAML tags that can be used to assert types. This is used in the workflow engine to distinguish
+// expressions from values unambiguously.
+package yaml //nolint:testpackage
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+)
+
+func assertEqualsYAML(t *testing.T, got Node, expected Node, path ...string) {
+	t.Helper()
+	if got.Tag() != expected.Tag() {
+		if len(path) > 0 {
+			t.Fatalf(
+				"Mismatching tag at %s, got %s, expected %s",
+				strings.Join(path, " -> "),
+				got.Tag(),
+				expected.Tag(),
+			)
+		}
+		t.Fatalf("Mismatching tag, got %s, expected %s", got.Tag(), expected.Tag())
+	}
+	if got.Type() != expected.Type() {
+		if len(path) > 0 {
+			t.Fatalf(
+				"Mismatching type at %s, got %s, expected %s",
+				strings.Join(path, " -> "),
+				got.Type(),
+				expected.Type(),
+			)
+		}
+		t.Fatalf("Mismatching type, got %s, expected %s", got.Type(), expected.Type())
+	}
+	if got.Value() != expected.Value() {
+		if len(path) > 0 {
+			t.Fatalf(
+				"Mismatching value at %s, got %s, expected %s",
+				strings.Join(path, " -> "),
+				got.Value(),
+				expected.Value(),
+			)
+		}
+		t.Fatalf("Mismatching value, got %s, expected %s", got.Value(), expected.Value())
+	}
+	gotContents := got.Contents()
+	expectedContents := expected.Contents()
+	if len(gotContents) != len(expectedContents) {
+		if len(path) > 0 {
+			t.Fatalf(
+				"Mismatching contents count at %s, got %d, expected %d",
+				strings.Join(path, " -> "),
+				len(gotContents),
+				len(expectedContents),
+			)
+		}
+		t.Fatalf("Mismatching value, got %d, expected %d", len(gotContents), len(expectedContents))
```

Comment:
Taking advantage of knowing the lengths are the reason for the mismatch, I would find it useful to say something about the lengths being different in the message.